controllers/awsmachine_controller.go

```diff
@@ -448,57 +448,54 @@ func (r *AWSMachineReconciler) getOrCreate(scope *scope.MachineScope, ec2svc ser
 		return nil, err
 	}
 
-	if instance == nil {
-		scope.Info("Creating EC2 instance")
-		// Create a new AWSMachine instance if we couldn't find a running instance.
+	// If we find an instance, return it
+	if instance != nil {
+		return instance, nil
+	}
+	// Otherwise create a new instance
+	scope.Info("Creating EC2 instance")
+
+	userData, err := scope.GetRawBootstrapData()
+	if err != nil {
+		r.Recorder.Eventf(scope.AWSMachine, corev1.EventTypeWarning, "FailedGetBootstrapData", err.Error())
+		return nil, err
+	}
 
-		userData, err := scope.GetRawBootstrapData()
+	if scope.UseSecretsManager() {
+		compressedUserData, err := userdata.GzipBytes(userData)
 		if err != nil {
-			r.Recorder.Eventf(scope.AWSMachine, corev1.EventTypeWarning, "FailedGetBootstrapData", err.Error())
 			return nil, err
 		}
-
-		if scope.UseSecretsManager() {
-			compressedUserData, err := userdata.GzipBytes(userData)
-			if err != nil {
-				return nil, err
-			}
-			// Do an initial check in case manager was terminated prematurely
-			if scope.GetSecretPrefix() == "" {
-				prefix, chunks, err := secretSvc.Create(scope, compressedUserData)
-				if err := scope.SetSecretPrefix(prefix); err != nil {
-					return nil, err
-				}
-				if err := scope.SetSecretCount(chunks); err != nil {
-					return nil, err
-				}
-				if err != nil {
-					r.Recorder.Eventf(scope.AWSMachine, corev1.EventTypeWarning, "FailedCreateAWSSecretsManagerSecrets", err.Error())
-					scope.Error(err, "Failed to create AWS Secret entry", "secretPrefix", prefix)
-					delErr := secretSvc.Delete(scope)
-					if delErr != nil {
-						scope.Error(delErr, "Failed to clean up AWS Secret entries", "secretPrefix", prefix)
-						return nil, delErr
-					}
-					return nil, err
-				}
-			}
-			// Register the Secret ARN immediately to avoid orphaning AWS resources on delete
-			if err := scope.PatchObject(); err != nil {
+		prefix, chunks, serviceErr := secretSvc.Create(scope, compressedUserData)
+		// Only persist the AWS Secrets Manager entries if there is at least one
+		if chunks > 0 {
+			if err := scope.SetSecretPrefix(prefix); err != nil {
 				return nil, err
 			}
-			encryptedCloudInit, err := secretsmanager.GenerateCloudInitMIMEDocument(scope.GetSecretPrefix(), scope.GetSecretCount(), scope.AWSCluster.Spec.Region)
-			if err != nil {
-				r.Recorder.Eventf(scope.AWSMachine, corev1.EventTypeWarning, "FailedGenerateAWSSecretsManagerCloudInit", err.Error())
+			if err := scope.SetSecretCount(chunks); err != nil {
 				return nil, err
 			}
-			userData = encryptedCloudInit
 		}
-
-		instance, err = ec2svc.CreateInstance(scope, userData)
+		// Register the Secret ARN immediately to avoid orphaning whatever AWS resources have been created
```

Comment:
possibly follow up work, but I wonder if it would be worth it to compute the secretprefix and count prior to creating the secrets for the purposes of persisting this info prior to making the call.

We could potentially hit a situation where we haven't created the secrets yet, but recorded the fields, but that would remove any scenario of orphaning.

Reply:
Possibly. This would be in the case that the controller failed to persist the data in between creating the secret and patching the resource?

Reply:
Correct, but it's no worse than it would be for other resources that can't be tagged on creation :)